src/include/homestore/index/index_internal.hpp

```diff
@@ -63,31 +63,61 @@ enum class index_buf_state_t : uint8_t {
 };
 
 ///////////////////////// Btree Node and Buffer Portion //////////////////////////
-class IndexBuffer;
+struct IndexBuffer;
+struct IndexBufferGroup;
 typedef std::shared_ptr< IndexBuffer > IndexBufferPtr;
+typedef std::shared_ptr< IndexBufferGroup > IndexBufferGroupPtr;
 
 struct IndexBuffer {
     uint8_t* m_node_buf{nullptr};                            // Actual buffer
     index_buf_state_t m_buf_state{index_buf_state_t::CLEAN}; // Is buffer yet to persist?
     BlkId m_blkid;                                           // BlkId where this needs to be persisted
-    IndexBufferPtr m_next_buffer{nullptr};                   // Next buffer in the chain
-    // Number of leader buffers we are waiting for before we write this buffer
-    sisl::atomic_counter< int > m_wait_for_leaders{0};
+    IndexBufferGroupPtr m_buf_group;                         // Buffer group to which it belongs.
 
     IndexBuffer(BlkId blkid, uint32_t buf_size, uint32_t align_size);
+    ~IndexBuffer();
 
     BlkId blkid() const { return m_blkid; }
     uint8_t* raw_buffer() { return m_node_buf; }
 
     bool is_clean() const { return (m_buf_state == index_buf_state_t::CLEAN); }
+    std::string to_string() const {
+        return fmt::format("IndexBuffer {} blkid={} state={} node_buf={}",
+                           reinterpret_cast< void* >(const_cast< IndexBuffer* >(this)), m_blkid.to_integer(),
+                           static_cast< int >(m_buf_state), static_cast< void* >(m_node_buf));
+    }
+};
+
+struct IndexBufferGroup {
+    // First and second index buffers are flushed in order.
+    std::array< IndexBufferPtr, 2 > m_bufs;
+    IndexBufferGroupPtr m_next_group{nullptr}; // Next buffer group in the chain
+    // Number of leader group buffers we are waiting for before we write this buffer
+    sisl::atomic_counter< int > m_wait_for_leaders{0};
+    int32_t m_cur_indx{0};
+
+    IndexBufferGroup(IndexBufferPtr first, IndexBufferPtr second = nullptr) : m_bufs{first, second} {}
+    IndexBufferPtr first() { return m_bufs[0]; }
+    IndexBufferPtr second() { return m_bufs[1]; }
+    int count() { return m_bufs[1] ? 2 : 1; }
+
+    std::string to_string() const {
+        return fmt::format(
+            "IndexBufferGroup {} m_bufs = [{}:{}, {}] next_group={} wait_for_leaders={}",
+            reinterpret_cast< void* >(const_cast< IndexBufferGroup* >(this)), static_cast< void* >(m_bufs[0].get()),
+            m_bufs[0]->blkid().to_integer(),
+            m_bufs[1] ? fmt::format("{}:{}", static_cast< void* >(m_bufs[1].get()), m_bufs[1]->blkid().to_integer())
+                      : "null",
+            static_cast< void* >(m_next_group.get()), m_wait_for_leaders.get());
```

Comment:
Not mandatory: we do have voidptr_cast (for shorter words - if you prefer)

Reply:
Ack